E.P.I.C (Extended PID Configuration)

In [1]:
#Import libraries
import pandas as pd
import numpy as np
#Cambia todo el formato de pandas . no recomendable
#pd.options.display.float_format = '{:,.3f}'.format


#Python Collections (Arrays)

#There are four collection data types in the Python programming language:

 #   List is a collection which is ordered and changeable. Allows duplicate members.
 #   Tuple is a collection which is ordered and unchangeable. Allows duplicate members.
 #   Set is a collection which is unordered, unchangeable*, and unindexed. No duplicate members.
 #   Dictionary is a collection which is ordered** and changeable. No duplicate members. 
 #   When we say that dictionaries are ordered, it means that the items have a defined order, and that order will not change.
 #   Unordered means that the items does not have a defined order, you cannot refer to an item by using an index.

In [2]:
pwd

'c:\\Users\\automation\\Desktop\\DataAnalisisPID'

In [3]:
#Load PID data into a dataframe
#In Sublime text, select "Saving with encoding: utf-8"
#https://stackoverflow.com/questions/18171739/unicodedecodeerror-when-reading-csv-file-in-pandas-with-python
#Delete last empty row of the csv file.
df_raw_data = pd.read_csv('PID.csv', encoding = "utf-8", engine='python')

In [4]:
#Print data raw
df_raw_data

ShapeKey Prop_PID_Element  Prop_PID_Number
0       16872                W             1.00
1       16944                W             2.00
2       16125               XA             1.10
3       16126               SC             1.10
4       14434                M             1.10
..        ...              ...              ...
299     34799               XA             2.34
300     34812               XA             2.35
301     34813                P             2.35
302     18619             SIA±             2.82
303     15757                L             2.99

[304 rows x 3 columns]

In [5]:
#datatypes of each column
df_raw_data.dtypes

ShapeKey              int64
Prop_PID_Element     object
Prop_PID_Number     float64
dtype: object

In [6]:
#3 decimals for PID
df_raw_data['Prop_PID_Number'] = df_raw_data['Prop_PID_Number'].apply(lambda x: f'{x:.3f}')

In [7]:
#datatypes of each column
df_raw_data.dtypes

ShapeKey             int64
Prop_PID_Element    object
Prop_PID_Number     object
dtype: object

In [8]:
#Print data raw
df_raw_data

ShapeKey Prop_PID_Element Prop_PID_Number
0       16872                W           1.000
1       16944                W           2.000
2       16125               XA           1.100
3       16126               SC           1.100
4       14434                M           1.100
..        ...              ...             ...
299     34799               XA           2.340
300     34812               XA           2.350
301     34813                P           2.350
302     18619             SIA±           2.820
303     15757                L           2.990

[304 rows x 3 columns]

In [9]:
#Show headers
df_raw_data.columns

Index(['ShapeKey', 'Prop_PID_Element', 'Prop_PID_Number'], dtype='object')

In [10]:
#Filter data (#Select 2 columns using [[]])
df_filtered_data = df_raw_data[['Prop_PID_Number','Prop_PID_Element']]

In [11]:
#print filtered data
df_filtered_data

Prop_PID_Number Prop_PID_Element
0             1.000                W
1             2.000                W
2             1.100               XA
3             1.100               SC
4             1.100                M
..              ...              ...
299           2.340               XA
300           2.350               XA
301           2.350                P
302           2.820             SIA±
303           2.990                L

[304 rows x 2 columns]

In [12]:
#Find duplicates values

s = df_filtered_data['Prop_PID_Number'].duplicated()
#print(s)
#s.size()
duplicated_list = s.tolist()

for items in duplicated_list:
    while
        print(items)


SyntaxError: invalid syntax (Temp/ipykernel_5268/616192224.py, line 9)

In [ ]:
#datatypes of each column
df_filtered_data.dtypes

Find uniques values for PID and Elements
Create a database with PID and which elements contain:
#List items are ordered, changeable, and allow duplicate values.
pid_list = [
    {'PID':'1.100','Description':'Intake air fan','Elements':{'XA':True,'SC':True,'M':'True','V':True,'PDI':False,'TSA+':False,'F':False,'C':False},'SW_module':'S_MOTOR1'},
    {'PID':'1.160','Description':'Intake air fan','Elements':{'XA':True,'SC':True,'M':'True','V':True,'PDI':False,'TSA+':False,'F':False,'C':False},'SW_module':'S_MOTOR1'},
    {'PID':'1.150','Description':'Exhaust air fan','Elements':{'XA':False,'SC':True,'M':'True','V':False,'PDI':False,'TSA+':True,'F':False,'C':False},'SW_module':'S_MOTOR1'},
    {'PID':'1.200','Description':'Recirculation air fan','Elements':{'XA':True,'SC':True,'M':'True','V':True,'PDI':False,'TSA+':False,'F':False,'C':False},'SW_module':'S_MOTOR1'},
    {'PID':'1.210','Description':'Recirculation air fan','Elements':{'XA':True,'SC':True,'M':'True','V':True,'PDI':False,'TSA+':False,'F':False,'C':False},'SW_module':'S_MOTOR1'},
    {'PID':'1.218','Description':'Differencial presure','Elements':{'XA':False,'SC':False,'M':'False','V':False,'PDI':True,'TSA+':False,'F':False,'C':False},'SW_module':'S_LIMIT_HI_LO'}
]

In [ ]:
#PID number to list
pid_list = list(df_filtered_data['Prop_PID_Number'])
pid_list.sort()
print(pid_list)

In [ ]:
len(pid_list)

In [ ]:
#Remove duplicates
pid_list = list(dict.fromkeys(pid_list))
print(pid_list)

In [ ]:
len(pid_list)

In [ ]:
#Remove unexistent PIDs
#Improve using regular expresions
pid_list.remove("1.000") #as example
print(pid_list)

In [ ]:
#Elements to list
element_list = list(df_filtered_data['Prop_PID_Element'])
element_list.sort()
print(element_list)

In [ ]:
len(element_list)

In [ ]:
#Remove duplicates
element_list = list(dict.fromkeys(element_list))
print(element_list)

In [ ]:
len(element_list)

In [ ]:
#Create a datatype to save data of pid
from dataclasses import dataclass, field

@dataclass
class PID:
    number: str  = "X.XXX"
    description : str = "Description"



In [ ]:
#Try to create a list of dataclases
epic_list = []
for pids_numbers in pid_list:
    epic_list.append( PID(pids_numbers,"Description") )

In [ ]:
#print(epic_list)

for obj in epic_list:
    print(obj.number, obj.description)
